### PR TITLE
sysfs: (fixes #26) Add GPIO support for the Jetson Orin AGX board

### LINKED ISSUE #26: Support GPIO pins on Jetson Orin boards
**What kind of new feature are you looking for?**
- Hardware: Jetson Orin AGX (and maybe other Jetson Orin boards, and maybe even other Jetson boards?)

**Do you plan to:**
- Contribute an initial driver: Unsure: I'm likely to fork this repo and play around with things, but no guarantees I'll actually get anywhere or send a PR back when I'm finished
- Write unit tests: No: the real proof that things work is in running on the proper hardware, and the github action runners for this repo don't have a Jetson Orin board connected.
- Update https://github.com/periph/cmd to use the new functionality: Probably not: my first guess is that it probably doesn't need updating, and my second guess is that whoever contributes the initial driver should update this as well.

**Description of the problem**
For most sysfs systems, you can enable a GPIO pin by writing a number into `/sys/class/gpio/export`, at which point you get a new symlink named `/sys/class/gpio/gpioN` (where `N` is the number you wrote). This is so universally common that it's hard-coded into this repo, at https://github.com/periph/host/blob/main/sysfs/gpio.go#L493. 

Unfortunately, the [Jetson Orin](https://www.nvidia.com/en-us/autonomous-machines/embedded-systems/jetson-orin/) (at least, the AGX model, not sure about the others) doesn't name their symlinks `gpioN`. For example, if you `cd /sys/class/gpio; echo "336" > export`, instead of getting a symlink named `gpio336`, you get one named `PDD.00`. However, the gpio.go file in this repo hardcodes that the symlink should always be at `gpio336` (or another number), and then we get "file not found"/"pin not exported" errors later if we try to use any of the GPIO pins. 

To support the Jetson Orin AGX (and possibly other boards in the Jetson Orin family, though I haven't tried any of them yet), please make this configurable. The line that needs to change is https://github.com/periph/host/blob/main/sysfs/gpio.go#L493, though various other things probably need to change along with it (maybe the pin definitions need a new field? but that sounds like a lot of work, and hopefully there's an easier approach I haven't thought of). 

I'd be happy to help brainstorm ideas of what to do, if that would help. I'm on the fence for whether I'm also willing to make the changes myself, but I'm not a flat-out "no" on that end.

### PR DESCRIPTION
**Pull request by penguinland**
This is a fix for https://github.com/periph/host/issues/26. It's ugly, and I kinda suspect I should have put the data elsewhere, but it works. I'm totally open to suggestions of moving things elsewhere, though!

I am relatively new to both Go and sysfs. I've probably written something unidiomatic, and would be happy to change whatever it is.

The implementation:
- At initialization, store the board name (the contents of `/proc/device-tree/model`). 
- When you need to generate the `root` directory of a GPIO pin, check the board name: if it's a Jetson Orin AGX, use a giant lookup table to go from pin number to their custom name, and for literally everything else ever, use the standard name like usual.

Tried on a Jetson Orin AGX on my desk: I'm able to turn the GPIO pins on and off! I have not tried it on other devices, but I believe this repo has a CI system that will test it on various other hardware, and no changes on other hardware are intended.

> 3. Once integrated, send a PR to https://github.com/periph/cmd to leverage the new functionality (if relevant).

I don't know if that's relevant or what would go in the PR. Any advice?

### CODE REVIEW

File: sysfs/gpio.go [hunk 1]
```diff
@@ -505,7 +505,75 @@ func (d *driverGPIO) parseGPIOChip(path string) error {
...
+		// For unknown pins, try the default name but expect failures regardless.
+	}
+	// Nearly all boards use this naming scheme:
+	return fmt.Sprintf("/sys/class/gpio/gpio%d/", pinNumber)
+}
+
+var boardName string // The contents of /proc/device-tree/model, if it exists
+
+func initBoardName() {
+	f, err := fileIOOpen("/proc/device-tree/model", os.O_RDONLY)
```
Comment (maruel) [inline sysfs/gpio.go:562]: Let's use https://pkg.go.dev/periph.io/x/host/v3/distro#DTModel

File: sysfs/gpio.go [hunk 2]
```diff
@@ -490,7 +490,7 @@ func (d *driverGPIO) parseGPIOChip(path string) error {
 		p := &Pin{
```
Comment (maruel) [inline sysfs/gpio.go:490]: Query the board before the loop here, instead of using a global variable.

File: sysfs/gpio.go [hunk 3]
```diff
@@ -505,7 +505,75 @@ func (d *driverGPIO) parseGPIOChip(path string) error {
 	return nil
 }
 
+func getSymlinkRoot(pinNumber int) string {
+	// The NVidia Jetson Orin AGX uses nonstandard names within /sys/class/gpio. This is a mapping
+	// from pin numbers to their names on that machine.
+	jetsonOrinAgxPins := map[int]string{
+		316: "AA.00", 317: "AA.01", 318: "AA.02", 319: "AA.03", 320: "AA.04", 321: "AA.05",
```
Comment (maruel) [inline sysfs/gpio.go:512]: This code allocates the whole map at every single function call. This is inefficient. A slightly better version would look like this (simplified form):

```
var mapping[...]string{"AA.00", "AA.01", ...}

p -= 316
return mapping[i]
```

I have concerns though;

https://docs.nvidia.com/jetson/archives/r35.1/DeveloperGuide/text/HR/JetsonModuleAdaptationAndBringUp/JetsonAgxOrinSeries.html#changing-the-pinmux

and

https://docs.nvidia.com/jetson/archives/r35.1/DeveloperGuide/text/HR/JetsonModuleAdaptationAndBringUp/JetsonAgxOrinSeries.html#identifying-the-gpio-number

has two interesting bits:

- In the JetPack 5.0.2-GA and later, you can use the GPIO sysfs tool to update the pinmux register
- Because the Jetson module dynamically registers GPIOs, search the kernel messages to check the GPIO allocation ranges for each GPIO group

so basically, they say the pin mapping name can change?

**Review by maruel — COMMENTED** [review]
Thanks for the contribution.

File: sysfs/gpio.go (hunk 3 above)
Comment (penguinland) [reply sysfs/gpio.go:512]: > the pin mapping name can change

Ugh... you're right, but fixing that so it Just Works for everyone seems very hard. To search the kernel messages, I ran `dmesg | grep gpiochip` on my Orin (per your second link), and discovered that I need to use `sudo` to use `dmesg`. but I really don't want to require root access to use this library, so the approach of "parse `sudo dmesg | grep gpio` to get the base numbers" isn't going to work in a straightforward way.

How would you feel about just putting a warning in the readme, that if you change the pinmux on an Orin, this library probably won't work right? My gut instinct is that almost no one will change the pinmux, and supporting that feature isn't worth the hassle, though I could be wrong. and getting something that works most of the time beats something that doesn't work at all, even if it's not perfect.


(I'm still working on the rest of your feedback, but should be done soon)

**Comment by maruel** [comment]
I was wondering, could we have the code look at gpiochipXX symlinks, then look at /base and /label and try to calculate from there?

**Comment by maruel** [comment]
https://www.kernel.org/doc/html/latest/driver-api/gpio/index.html?highlight=names#c.gpio_chip tells me `names` is probably filled up with the right data.

**Comment by penguinland** [comment]
You might be right that there's a way to find the right names even for dynamically renumbered pins by examining that data structure. but I feel out of my depth (where would I even get that data structure from?), and don't have the time right now to research this further. I've gotten far enough that my main work is no longer blocked, so I'm going to go back to that instead. You're welcome to modify this PR further to support those extra cases (or close the PR, or something else). 

Thank you for the advice and feedback you've provided!

**Comment by penguinland** [comment]
Closing because no one (including myself) has looked at this for 1.5 years.
